### DISCUSSION TITLE
Main bugs bothering us at the moment

### DISCUSSION THREAD

**Discussion by ffMathy** [vitest-dev/vscode#538] [Feedback]
I am filing this as a discussion, as I don't believe the current repo structure is very friendly towards examples where a repro can't be given. This is one of those cases. But I still think you guys deserve to know of the instability in the extension that has been bothering us since the rewrite of the extension.

You guys are doing a phenomenal job (that's also why I am sponsoring), and it's amazing to see almost daily releases of the extension sometimes. So this is not meant as a rant. It's meant to be constructive.

Here are the main issues. Again, I'd love to provide a repro, but haven't been able to for the past many months. Either due to time pressure (Q4 goals coming up), or a complex setup causing it to take too many company resources to do so.

I know that you're thinking "well, put your money where your mouth is", and companies are generally not very understanding of contributing to open source. That sucks. As an open source contributor myself, I understand this. However, you may still have interest in fixing this, since I am assuming it affects a lot of other users too.

# Timer not resetting
We often get in a situation where the timer just keeps running. This happens typically if we press "refresh" while a test is running, or if we click "stop" while a test is running.

<img width="623" alt="image" src="https://github.com/user-attachments/assets/5f71d724-65bc-468f-836a-4b1685dd95a0">

It makes me wonder if `finally` clauses are being used correctly to properly dispose resources in the event of failures or other scenarios?

# Stop has to be pressed twice
Often, when running a long-running test, pressing stop once isn't enough. At the second press of "stop", the test truly stops.

# Race conditions while tests are being discovered
While tests are being discovered, if you try to run a test, the whole extension goes into a limbo where you no longer can run anything until you've refreshed all tests again. So we've developed a habbit of waiting a lot of seconds for the discovery to be done.

# Running a single test within a test file marks all other tests as "skipped" in the UI
This is annoying, because often you want to still be able to see if the other tests passed or failed the last time.

# Sometimes Vitest just gets in a weird state
This typically happens if I have just been debugging a test and then click "stop" in the debugger. After this, I can't start tests anymore normally (only in debugging) and have to click "refresh".

**Comment by sheremet-va** [comment]
> Often, when running a long-running test, pressing stop once isn't enough. At the second press of "stop", the test truly stops.

This is intentional. Vitest awaits when the test gracefully completes at first. If you have a long running test, it will wait until it's completed (in case you need to do cleanup) - this is also how the CLI works

> Running a single test within a test file marks all other tests as "skipped" in the UI

This can be fixed no problem 

> While tests are being discovered, if you try to run a test, the whole extension goes into a limbo where you no longer can run anything until you've refreshed all tests again. So we've developed a habbit of waiting a lot of seconds for the discovery to be done.

Interesting. This should be possible to fix without issues

**Reply by ffMathy** [reply]
Awesome - you rock 💪 

About the timer, is that easy to fix too?

**Reply by sheremet-va** [reply]
> Awesome - you rock 💪 
> 
> 
> 
> About the timer, is that easy to fix too?

I know why it happens (promise is hanging), but I don't know how it happens (what triggers it). I fixed already some cases, but looks like there are more.

**Reply by sheremet-va** [reply]
By the way, are you using the latest extension version?

**Reply by ffMathy** [reply]
Yes. We use it in a DevContainer, and that auto-updates the extension every day on startup for all of our developers in the team.

**Reply by ffMathy** [reply]
Just checked. Right now it's `v1.8.0`.

**Reply by sheremet-va** [reply]
> Interesting. This should be possible to fix without issues

I can't reproduce this. I need a more detailed explanation what you mean by discovery and how you can run a test during test discovery

**Reply by ffMathy** [reply]
By "discovery" I mean the phase where it expands every single test file in the test explorer to show the tests beneath it. This takes around 1 second for each test file for us.

A "discovery" (in my terms) can be triggered by hitting refresh.

During that time, if it hasn't completed this for the full project and we run a test, it goes in limbo.

**Reply by sheremet-va** [reply]
> A "discovery" (in my terms) can be triggered by hitting refresh.

So, what you are saying is that you had some tests expanded in the test tree, and then you refreshed Vitest or the VSCode and while it was collecting tests you cannot start new ones?

**Reply by sheremet-va** [reply]
Another question - I added `vitest.experimentalStaticAstCollect` basically just for you because your project takes a lot of time to load tests. Did you enable it?

**Reply by ffMathy** [reply]
I tried enabling it at some point, but it was failing with some error, so I disabled it again. Do you want that error?

Although I want to highlight that the wait time is not a problem. The thing that is the problem is that we sometimes accidentally click "start" because we believe it's done discovering. But then that brings it into limbo, and we have to click "refresh" again.

**Reply by sheremet-va** [reply]
> The thing that is the problem is that we sometimes accidentally click "start" because we believe it's done discovering.

Vitest runs files to discover tests. While it runs those files, it cannot run other tests in parallel. I am not sure it puts it into a limbo, I think you just have to wait until all tests are collected. I also have a project that takes 1-2s to collect tests and I cannot reproduce the limbo situation yet.

If we separate test discovery from running, then we don't need to dance different promises around eliminating the risk of a limbo.

 > I tried enabling it at some point, but it was failing with some error, so I disabled it again. Do you want that error?

Yes, I am planning to enable it by default.

**Reply by ffMathy** [reply]
Separating the test discovery from the running would be great. Is that planned currently?

Will get back to you with the error I am getting soon. I need to wrap up some PRs before I go on weekend first. Might be next week.

**Reply by sheremet-va** [reply]
> Separating the test discovery from the running would be great. Is that planned currently?

`vitest.experimentalStaticAstCollect` is how I plan to separate them

**Reply by ffMathy** [reply]
Aha. I see.

I get these errors now:

```
[INFO 10:41:38 AM] [Worker] Skipping withHttpRecordings (unknown call)
[Error 10:41:38 AM] Error in /workspaces/WATS/src/apps/REDACTED/src/domain/actions/REDACTED.test.ts No test suite found in file /workspaces/WATS/src/apps/REDACTED/src/domain/actions/REDACTED.test.ts
[Error 10:41:38 AM] Error in /workspaces/WATS/src/apps/REDACTED/src/domain/actions/REDACTED.test.ts No test suite found in file /workspaces/WATS/src/apps/REDACTED/src/domain/actions/REDACTED.test.ts
[Error 10:41:38 AM] Error in /workspaces/WATS/src/apps/REDACTED/src/domain/actions/REDACTED.test.ts No test suite found in file /workspaces/WATS/src/apps/REDACTED/src/domain/actions/REDACTED.test.ts
[Error 10:41:38 AM] Error in /workspaces/WATS/src/apps/REDACTED/src/domain/update-hooks/REDACTED.test.ts No test suite found in file /workspaces/WATS/src/apps/REDACTED/src/domain/update-hooks/REDACTED.test.ts
[Error 10:41:38 AM] Error in /workspaces/WATS/src/apps/REDACTED/src/domain/update-hooks/REDACTED.test.ts No test suite found in file /workspaces/WATS/src/apps/REDACTED/src/domain/update-hooks/REDACTED.test.ts
[Error 10:41:38 AM] Error in /workspaces/WATS/src/apps/REDACTED/src/domain/update-hooks/REDACTED.test.ts No test suite found in file /workspaces/WATS/src/apps/REDACTED/src/domain/update-hooks/REDACTED.test.ts
[Error 10:41:38 AM] Error in /workspaces/WATS/src/apps/wats-job-runner/src/domain/jobs/REDACTED/sync/REDACTED.test.ts
```

**Reply by sheremet-va** [reply]
> I get these errors now

Do you use `vitest.globals`?

**Reply by ffMathy** [reply]
Yup, I do 🙂 

Also, in regards to timings, I just encountered another bug. If you have a worker count higher than 1 and run all your tests, the timings are actually calculated wrong.

**Reply by sheremet-va** [reply]
The extension doesn't calculate overall time. VSCode does it by itself by summing all durations probably. We just report how long a single test took to run.

**Reply by sheremet-va** [reply]
Pushed 1.8.1. It should fix the issue with tests unintentionally marked as `skip`, and it should also fix an error with "No test suite found in the file" when `vitest.experimentalStaticAstCollect` is enabled. There is still a known issue with how `.each` tests are reported, I am trying to find a nice solution there

**Reply by ffMathy** [reply]
It works. And it's lightning fast. Thanks a lot. We don't have a lot of `.each` tests yet, but we do have a few. I'll keep this enabled either way, and eagerly await a fix. Nice work.
